### PR TITLE
Add type definitions for smtp-server.

### PR DESCRIPTION
**Pull request by markisme**
- [x] The package does not provide its own types.
- [x] match the npm package name.
- [x] Run `tsc` without errors.
- [x] Include the required [files](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package) and header. Base these on the README, *not* on an existing project.

### CODE REVIEW

File: smtp-server/package.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (andy-ms) [inline smtp-server/package.json]: Shouldn't need this file.

File: smtp-server/smtp-server.d.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+export interface Address {
+    address: string;
+    args: {
+        'SIZE'?: number,
+        'RET'?: string,
+        'NOTIFY'?: string
+    };
+}
+
+export interface SMTPCommands {
```
Comment (andy-ms) [inline smtp-server/smtp-server.d.ts]: An empty interface is equivalent to `{}`. Also, this doesn't seem to be used.

File: smtp-server/smtp-server.d.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    onAuth?: (auth: Authentication, session: Session, callback: (err?: Error, response?: AuthenticationResponse) => any) => any;
+    onConnect?: (session: Session, callback: (err?: Error) => any) => any;
+    onMailFrom?: (address: Address, session: Session, callback: (err?: Error) => any) => any;
+    onRcptTo?: (address: Address, session: Session, callback: (err?: Error) => any) => any;
+    onData?: (stream: Stream, session: Session, callback: (err?: Error) => any) => any;
+    onClose?: (session: Session) => any;
+}
+
+export interface Authentication {
+    method: string;//'PLAIN', 'LOGIN' or 'XOAUTH2'
```
Comment (andy-ms) [inline smtp-server/smtp-server.d.ts]: You can use a string union: `"PLAIN" | "LOGIN" | "XOAUTH2"`

File: smtp-server/smtp-server.d.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    onMailFrom(address: Address, session: Session, callback: (err?: Error) => any): any;
+    onRcptTo(address: Address, session: Session, callback: (err?: Error) => any): any;
+}
+
+export interface SMTPServerOptions {
+    secure?: boolean; //default is false.
+    key?: Buffer;
+    cert?: Buffer;
+    ca?: Buffer;
+    name?: string;//optional hostname of the server.
```
Comment (andy-ms) [inline smtp-server/smtp-server.d.ts]: Why not use a JSDoc comment for this? `/** Optional hostname of the server. */` on a line above it. JSDoc comments are used for an editor to provide info on hover, while this comment will be ignored.

File: smtp-server/smtp-server.d.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
+// Type definitions for smtp-server v1.15.0
+// Project: https://github.com/andris9/smtp-server
+// Definitions by: markisme <https://github.com/markisme/>
+// Definitions: https://github.com/markisme/smtp-server.git
```
Comment (andy-ms) [inline smtp-server/smtp-server.d.ts]: This should be `// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped`.

File: smtp-server/smtp-server.d.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+export interface Envelope {
+    mailFrom: Address;
+    rcptTo: Address[];
+}
+
+export interface Address {
+    address: string;
+    args: {
+        'SIZE'?: number,
```
Comment (andy-ms) [inline smtp-server/smtp-server.d.ts]: Quotes should be unnecessary here.

**Review by ghost — CHANGES_REQUESTED** [review]
Needs a `tsconfig.json`. See the [README](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package).

**Comment by andy-ms** [comment]
@markisme Please rebase onto the latest `types-2.0` to fix tests.

**Comment by markisme** [comment]
@andy-ms thanks very much, i will fix that as soon.😁

**Comment by markisme** [comment]
Maybe fixed.
Checklist

- [x] is correct [naming convention](http://definitelytyped.org/guides/contributing.html#naming-the-file)?

  - https://www.npmjs.com/package/smtp-server
  - http://bower.io/search/?q=smtp-server
  - others?

 

- [x] has a test file? (smtp-server-tests.ts or smtp-server-tests.tsx)
- [x] pass the Travis CI test?

**Comment by andy-ms** [comment]
Could you re-target this for `types-2.0` instead of `master`?

**Comment by markisme** [comment]
@andy-ms Did I need to reopen a PR on `types-2.0`?

**Comment by sandersn** [comment]
Yes, unless

**Pull request closed without merging**